<a href="https://colab.research.google.com/github/hibahrehman25-lang/ML_inter_Task1/blob/main/w03_data_contract.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-04 — Search Intelligence Data Contract

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/hibahrehman25-lang/ML_inter_Task1/blob/main/work/notebooks/w03_data_contract.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

In [ ]:
!pip -q install duckdb huggingface_hub

In [ ]:
from google.colab import userdata

HF_TOKEN = userdata.get("HF_TOKEN")

In [ ]:
con.sql("""
SELECT *
FROM 'hf://datasets/FlyRank/internship-warehouse/dim_content.parquet'
LIMIT 5
""").df()

,client_hash_id,content_hash_id,keyword_hash_id,url_hash_id,keyword_char_count,keyword_token_count,url_char_count,content_created_date,content_updated_date,content_type,...,category_count,keyword_created_date,provider_used,model_used,char_count,word_count,last_optimized_date,optimization_eligible_date,is_published,is_deleted
0,client_04660893ae39614a,content_004de9653278b5a4,keyword_e754999ab88dd9f2,url_d6091f18cf628794,22,4,108,2026-05-30,2026-07-01,keyword article,...,3,2026-05-12,gemini-generate-content,gemini-3-flash-preview,15682,2555,NaT,NaT,True,False
1,client_04660893ae39614a,content_00dc5efae381b2ab,keyword_4329d7aede8e208b,url_3a66d2f2e36823ca,31,6,95,2026-06-12,2026-07-01,keyword article,...,4,2026-06-01,gemini-generate-content,gemini-3-flash-preview,15438,2430,NaT,NaT,True,False
2,client_04660893ae39614a,content_01410f2556c327ac,keyword_9b08047d3d2a0406,url_809eda7a7e20b3b2,22,5,82,2026-05-09,2026-07-01,keyword article,...,4,2026-05-06,gemini-generate-content,gemini-3-flash-preview,16576,2645,NaT,NaT,True,False
3,client_04660893ae39614a,content_019f27f634053ca7,keyword_e7cec7ab1804c1c2,url_5fb42bafc4399861,14,3,92,2026-06-15,2026-06-15,keyword article,...,4,2026-06-01,gemini-generate-content,gemini-3-flash-preview,15457,2522,NaT,NaT,True,False
4,client_04660893ae39614a,content_01efa71faea45dcc,keyword_56b0062a1d8b7524,url_ece0abc3e5fb75f9,24,6,98,2026-05-21,2026-06-01,keyword article,...,4,2026-05-12,gemini-generate-content,gemini-3-flash-preview,15776,2552,NaT,NaT,True,False


In [ ]:
con.sql("""
DESCRIBE
SELECT *
FROM 'hf://datasets/FlyRank/internship-warehouse/dim_content.parquet'
""").df()

,column_name,column_type,null,key,default,extra
0,client_hash_id,VARCHAR,YES,None,None,None
1,content_hash_id,VARCHAR,YES,None,None,None
2,keyword_hash_id,VARCHAR,YES,None,None,None
3,url_hash_id,VARCHAR,YES,None,None,None
4,keyword_char_count,BIGINT,YES,None,None,None
5,keyword_token_count,BIGINT,YES,None,None,None
6,url_char_count,BIGINT,YES,None,None,None
7,content_created_date,DATE,YES,None,None,None
8,content_updated_date,DATE,YES,None,None,None
9,content_type,VARCHAR,YES,None,None,None


In [ ]:
from huggingface_hub import list_repo_files

files = list_repo_files(
    repo_id="FlyRank/internship-warehouse",
    repo_type="dataset",
    token=HF_TOKEN
)

for f in files:
    print(f)

.gitattributes
README.md
dim_clients.parquet
dim_content.parquet
fact_content_daily_performance/month=2025-01/data_0.parquet
fact_content_daily_performance/month=2025-02/data_0.parquet
fact_content_daily_performance/month=2025-03/data_0.parquet
fact_content_daily_performance/month=2025-04/data_0.parquet
fact_content_daily_performance/month=2025-05/data_0.parquet
fact_content_daily_performance/month=2025-06/data_0.parquet
fact_content_daily_performance/month=2025-07/data_0.parquet
fact_content_daily_performance/month=2025-08/data_0.parquet
fact_content_daily_performance/month=2025-09/data_0.parquet
fact_content_daily_performance/month=2025-10/data_0.parquet
fact_content_daily_performance/month=2025-11/data_0.parquet
fact_content_daily_performance/month=2025-12/data_0.parquet
fact_content_daily_performance/month=2026-01/data_0.parquet
fact_content_daily_performance/month=2026-02/data_0.parquet
fact_content_daily_performance/month=2026-03/data_0.parquet
fact_content_daily_performance/mont

In [ ]:
con.sql("""
DESCRIBE
SELECT *
FROM 'hf://datasets/FlyRank/internship-warehouse/fact_content_daily_performance/month=2026-03/data_0.parquet'
""").df()

,column_name,column_type,null,key,default,extra
0,report_date,DATE,YES,None,None,None
1,client_hash_id,VARCHAR,YES,None,None,None
2,content_hash_id,VARCHAR,YES,None,None,None
3,client_has_gsc,BOOLEAN,YES,None,None,None
4,client_has_ga4,BOOLEAN,YES,None,None,None
5,gsc_data_available,BOOLEAN,YES,None,None,None
6,ga4_data_available,BOOLEAN,YES,None,None,None
7,gsc_impressions,BIGINT,YES,None,None,None
8,gsc_clicks,BIGINT,YES,None,None,None
9,gsc_sum_position,BIGINT,YES,None,None,None


In [ ]:
con.sql("""
SELECT *
FROM 'hf://datasets/FlyRank/internship-warehouse/fact_content_daily_performance/month=2026-03/data_0.parquet'
LIMIT 5
""").df()

,report_date,client_hash_id,content_hash_id,client_has_gsc,client_has_ga4,gsc_data_available,ga4_data_available,gsc_impressions,gsc_clicks,gsc_sum_position,...,sessions_ai,ai_chatgpt,ai_perplexity,ai_gemini,ai_copilot,ai_claude,ai_meta,ai_other,scroll_events,month
0,2026-03-01,client_73cda7b4e4f265ea,content_b7e512995f79d5a6,True,False,True,<NA>,20,0,67,...,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,2026-03
1,2026-03-01,client_73cda7b4e4f265ea,content_05597932fe4da067,True,False,True,<NA>,1,0,0,...,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,2026-03
2,2026-03-01,client_73cda7b4e4f265ea,content_7a105f548d9c6916,True,False,True,<NA>,125,1,616,...,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,2026-03
3,2026-03-01,client_73cda7b4e4f265ea,content_905aa32a0230694e,True,False,True,<NA>,7,0,28,...,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,2026-03
4,2026-03-01,client_73cda7b4e4f265ea,content_a3ea9792f793ec72,True,False,True,<NA>,11,0,25,...,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,2026-03


In [ ]:
con.sql("""
SELECT
    COUNT(*) AS total_rows,
    COUNT(DISTINCT content_hash_id) AS unique_contents,
    COUNT(DISTINCT report_date) AS unique_dates
FROM 'hf://datasets/FlyRank/internship-warehouse/fact_content_daily_performance/month=2026-03/data_0.parquet'
""").df()

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,total_rows,unique_contents,unique_dates
0,9841378,331437,31


In [ ]:
con.sql("""
SELECT
    content_hash_id,
    COUNT(DISTINCT report_date) AS days_present
FROM 'hf://datasets/FlyRank/internship-warehouse/fact_content_daily_performance/month=2026-03/data_0.parquet'
GROUP BY content_hash_id
ORDER BY days_present DESC
LIMIT 10
""").df()

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,content_hash_id,days_present
0,content_05597932fe4da067,31
1,content_905aa32a0230694e,31
2,content_05434271b257bb68,31
3,content_22610b0934f8825e,31
4,content_712c365258cee05c,31
5,content_476c37c366920c1b,31
6,content_3dba50ae010f3f30,31
7,content_098eedbd77ec1de1,31
8,content_8935ed68eca88b01,31
9,content_fd377b1af29dcbb3,31


In [ ]:
con.sql("""
SELECT
    COUNT(*) AS total_rows,
    COUNT(*) FILTER (WHERE gsc_data_available IS TRUE) AS gsc_available_rows,
    COUNT(*) FILTER (WHERE ga4_data_available IS TRUE) AS ga4_available_rows
FROM 'hf://datasets/FlyRank/internship-warehouse/fact_content_daily_performance/month=2026-03/data_0.parquet'
""").df()

,total_rows,gsc_available_rows,ga4_available_rows
0,9841378,3611061,413966


In [ ]:
feature_df = con.sql("""
SELECT
    report_date,
    content_hash_id,
    gsc_impressions,
    gsc_clicks,
    gsc_avg_position,
    scroll_events,
    month
FROM 'hf://datasets/FlyRank/internship-warehouse/fact_content_daily_performance/month=2026-03/data_0.parquet'
WHERE gsc_data_available IS TRUE
LIMIT 10
""").df()

feature_df

,report_date,content_hash_id,gsc_impressions,gsc_clicks,gsc_avg_position,scroll_events,month
0,2026-03-01,content_b7e512995f79d5a6,20,0,3.350000,<NA>,2026-03
1,2026-03-01,content_05597932fe4da067,1,0,0.000000,<NA>,2026-03
2,2026-03-01,content_7a105f548d9c6916,125,1,4.928000,<NA>,2026-03
3,2026-03-01,content_905aa32a0230694e,7,0,4.000000,<NA>,2026-03
4,2026-03-01,content_a3ea9792f793ec72,11,0,2.272727,<NA>,2026-03
5,2026-03-01,content_36c36abc7650d7af,239,1,7.347280,<NA>,2026-03
6,2026-03-01,content_a7da352b73b02668,191,0,7.832461,<NA>,2026-03
7,2026-03-01,content_05434271b257bb68,55,0,3.272727,<NA>,2026-03
8,2026-03-01,content_d056587ff7faca0c,77,0,5.636364,<NA>,2026-03
9,2026-03-01,content_bfd1e41c2af250c8,2,0,4.500000,<NA>,2026-03


In [ ]:
con.sql("""
SELECT
    COUNT(*) AS total_rows,
    COUNT(gsc_impressions) AS impressions_not_null,
    COUNT(gsc_clicks) AS clicks_not_null,
    COUNT(gsc_avg_position) AS position_not_null,
    COUNT(scroll_events) AS scroll_not_null
FROM 'hf://datasets/FlyRank/internship-warehouse/fact_content_daily_performance/month=2026-03/data_0.parquet'
WHERE gsc_data_available IS TRUE
""").df()

,total_rows,impressions_not_null,clicks_not_null,position_not_null,scroll_not_null
0,3611061,3611061,3611061,3611061,2082695


In [ ]:
con.sql("""
SELECT
    gsc_impressions,
    gsc_clicks,
    gsc_avg_position
FROM 'hf://datasets/FlyRank/internship-warehouse/fact_content_daily_performance/month=2026-03/data_0.parquet'
WHERE gsc_data_available IS TRUE
LIMIT 10
""").df()

,gsc_impressions,gsc_clicks,gsc_avg_position
0,20,0,3.350000
1,1,0,0.000000
2,125,1,4.928000
3,7,0,4.000000
4,11,0,2.272727
5,239,1,7.347280
6,191,0,7.832461
7,55,0,3.272727
8,77,0,5.636364
9,2,0,4.500000


In [ ]:
con.sql("""
SELECT
    MIN(report_date) AS first_date,
    MAX(report_date) AS last_date
FROM 'hf://datasets/FlyRank/internship-warehouse/fact_content_daily_performance/month=2026-03/data_0.parquet'
""").df()

,first_date,last_date
0,2026-03-01,2026-03-31


In [ ]:
con.sql("""
SELECT
    content_hash_id,
    gsc_impressions,
    gsc_clicks,
    ROUND(
        CASE
            WHEN gsc_impressions > 0
            THEN (gsc_clicks * 100.0 / gsc_impressions)
            ELSE NULL
        END,2
    ) AS ctr_percent
FROM 'hf://datasets/FlyRank/internship-warehouse/fact_content_daily_performance/month=2026-03/data_0.parquet'
WHERE
    gsc_data_available IS TRUE
    AND gsc_impressions >= 100
ORDER BY ctr_percent ASC
LIMIT 10;
""").df()

,content_hash_id,gsc_impressions,gsc_clicks,ctr_percent
0,content_a7da352b73b02668,191,0,0.0
1,content_712c365258cee05c,223,0,0.0
2,content_8935ed68eca88b01,113,0,0.0
3,content_347a278bb1a646f5,134,0,0.0
4,content_4cd532e5e9edc02b,205,0,0.0
5,content_0c29bba4e77ab325,153,0,0.0
6,content_6cf70a5cb30e76f2,214,0,0.0
7,content_e833e04503f07324,215,0,0.0
8,content_952ce695b3f329e4,165,0,0.0
9,content_e48a4e8f6745ce7c,137,0,0.0


In [ ]:
con.sql("""
SELECT
    AVG(gsc_impressions) AS avg_impressions,
    AVG(gsc_clicks) AS avg_clicks,
    AVG(gsc_avg_position) AS avg_position
FROM 'hf://datasets/FlyRank/internship-warehouse/fact_content_daily_performance/month=2026-03/data_0.parquet'
WHERE gsc_data_available IS TRUE
""").df()

,avg_impressions,avg_clicks,avg_position
0,77.721642,0.227587,15.826651


## 1. Unit of analysis + time window

*One row = one what, over which dates? State it, then verify it below.*

## 1. Unit of Analysis + Time Window

### Unit of Analysis
One row represents the daily search performance of one content page (`content_hash_id`) for one client (`client_hash_id`) on one reporting date (`report_date`).

### Tables Used
- `fact_content_daily_performance`
- `dim_content` (used only if additional content metadata is needed)

### Time Window
This analysis uses the **March 2026** partition (`month = 2026-03`), covering data from **2026-03-01** to **2026-03-31**.

### Prediction / Ranking Objective
My selected lane is **CTR / Engagement Opportunity Scoring**. The goal is to rank content pages that receive high Google Search impressions but relatively few clicks so they can be prioritized for CTR optimization.

### Deliberately Excluded
I excluded AI referral metrics because they are outside the scope of my CTR opportunity task. I also excluded most GA4 metrics because they are available for only a subset of rows, making them unsuitable as primary features for this notebook.

## 2. Fields: feature / label / context / excluded

*Sort every field you plan to touch into these four buckets. Excluded needs a why.*

In [ ]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 2. Fields: Feature / Label / Context / Excluded

### Features
- gsc_impressions
- gsc_avg_position
- scroll_events
- client_has_gsc
- gsc_data_available

### Label / Proxy
Ranking content pages by **CTR improvement opportunity** (pages receiving high impressions but relatively low clicks).

### Context
- report_date
- client_hash_id
- content_hash_id
- month

### Excluded
- gsc_clicks (excluded from the final feature set because it is directly related to the ranking objective and may introduce leakage).
- AI referral metrics (not relevant to the selected lane).
- Most GA4 metrics (available for only a subset of rows).

## 3. Verify it with queries (grain, counts, missing values, windows)

*Every claim above gets a query cell here. A contract claim without a query next to it is a guess.*

In [ ]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 3. Verification Summary

The verification queries confirmed the following:

- The March 2026 partition contains **9,841,378** rows.
- It contains **331,437** unique content pages observed across **31** reporting dates.
- The selected time window spans from **2026-03-01** to **2026-03-31**.
- The grain is one content page per reporting date, verified by observing the same content page across multiple reporting dates.
- Google Search Console data is available for **3,611,061** rows, while Google Analytics data is available for **413,966** rows. Availability checks are therefore necessary before feature engineering.
- Pages with high impressions but very low CTR exist in the selected partition, supporting the CTR opportunity scoring objective.
- Within rows where GSC data is available, the average page received **77.72 impressions**, **0.23 clicks**, and had an average search position of **15.83**.

## Feature Notes

### gsc_impressions
Available before making optimization decisions because impressions have already been observed.

### gsc_avg_position
Available at the decision moment because the current search position is already known.

### scroll_events
Represents historical engagement. Missing values require appropriate handling before modeling.

### client_has_gsc
Known before analysis and indicates whether Google Search Console data exists for the client.

### gsc_data_available
Used to ensure only valid GSC observations are included in the analysis.

## Leakage Check

As an experiment, I considered using `gsc_clicks` as an input feature.

Because the objective is to identify pages with high impressions but relatively low clicks, including `gsc_clicks` would provide information that is directly related to the ranking objective and could artificially improve model performance.

To avoid leakage, I excluded `gsc_clicks` from the final feature set and retained only features that would be available independently at the decision moment.

## 4. Data limits

*What can this data never tell you? Unbalanced history, GSC-only early rows, window overlaps.*

In [ ]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 4. Data Limits

- Google Analytics (GA4) metrics are available for only a subset of rows, limiting analyses that depend on GA4 features.
- Scroll event data contains missing values and requires appropriate missing-value handling.
- This notebook analyzes only the March 2026 partition. Additional months should be evaluated before drawing broader conclusions.
- The analysis is intended to support decision-making by identifying CTR improvement opportunities. It does not establish causal relationships.

## Self-check

Before you submit, confirm each line honestly:

- [✔️] Every section above is filled — markdown thinking AND the code that backs it
- [✔️] The notebook runs top to bottom with no errors (Runtime → Run all)
- [✔️] No client names, URLs, or private queries anywhere
- [✔️] My claims use careful words: observed, measured, directional, decision-support
- [✔️] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.